Make it easier to change the encrypted config file for e2e tests

noahtallen opened the pull request:
### Changes proposed in this Pull Request
This PR makes it easier to swap config keys. As currently implemented, all existing builds will fail when the e2e config key is modified. The new system would allow existing branches/commits to decrypt the previous version of the config file while everyone gradually rebases with the new config file.

The new way to change the config will be:
1. Upload a new key to teamcity.
2. Modify the config file in VCS to add the key
3. Change `E2E_CONFIG_ENCRYPTION_KEY` to reference the new key param.

The previous builds will work, as the old key isn't yet deleted from TeamCity. Builds with the fresh encryption will work, as the new key is added to the newer config file.

If one needs to change all existing keys, or make it impossible for existing builds to read the config file, that's also easily done by modifying the all existing keys to match a new key.

### Testing instructions
- e2e should pass. (tbd, probably need to update the password in teamcity)

lynnbgriffith reviewed (approved):
👍 Thanks for looking into this.

noahtallen committed:
Encrypt again to double check

noahtallen committed:
Support multiple encryption keys

noahtallen commented on .teamcity/settings.kts:
```diff
@@ -64,6 +64,8 @@ project {
 		text("env.DOCKER_BUILDKIT", "1", label = "Enable Docker BuildKit", description = "Enables BuildKit (faster image generation). Values 0 or 1", allowEmpty = true)
 		password("CONFIG_E2E_ENCRYPTION_KEY", "credentialsJSON:819c139c-90a1-4803-8367-00e5aa5fdb07", display = ParameterDisplay.HIDDEN)
+		password("CONFIG_E2E_ENCRYPTION_KEY_JAN_22", "credentialsJSON:28f1cadd-35d5-44b1-bdca-4228fe2948b7", display = ParameterDisplay.HIDDEN)
+		param("E2E_CONFIG_ENCRYPTION_KEY", "%CONFIG_E2E_ENCRYPTION_KEY_JAN_22%")
```
This second value means that consumers can just reference a single config value for e2e keys, and that gets filled with the correct key based on vcs.

noahtallen commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -853,7 +853,7 @@ object QuarantinedE2ETests: BuildType( {
 
 				# Decrypt config
-				openssl aes-256-cbc -md sha1 -d -in ./config/encrypted.enc -out ./config/local-test.json -k "%CONFIG_E2E_ENCRYPTION_KEY%"
+				openssl aes-256-cbc -md sha1 -d -in ./config/encrypted.enc -out ./config/local-test.json -k "%E2E_CONFIG_ENCRYPTION_KEY%"
```
I renamed these so that previous code in vcs can access the older encryption key without changes. Once everyone is on the new system, one would only need to change the settings.kts file to update keys.

noahtallen commented:
cc @worldomonation @WunderBart 

We (myself and @griffbrad) were thinking that it would be helpful to allow rotating the config key without causing all existing builds to fail. (Which means everyone has to rebase, and it's hard to communicate that to everyone.) What do you think?

noahtallen commented:
I'm getting the classic "Last changes in TeamCity settings were not committed to VCS, use current settings from TeamCity server" error. I'll need to look into that further tomorrow. I'm surprised this didn't cause issues on previous updates to this file :thinking: Unless it normally doesn't work until merging?

noahtallen commented:
Very interesting! I reverted my changes in TeamCity, and now the build appears to work. I think this is what happens:

1. I upload a new password to TeamCity by changing the config.
2. I modify the config file in VCS
3. Since the TeamCity server changes were "not committed to VCS", TeamCity pulls from itself rather than VCS, not using our new changes here.
4. When using itself, it used its own config key, which is still outdated, to decrypt the new config file. **This fails the build.**
5. Then I reverted the TeamCity changes I made and re-triggered the build.
6. When running the build now, it thinks that the server changes are up to date with VCS, so it pulls from VCS.
7. The build succeeds now, because it accesses the new config key for the new config file.

Some more details:
- Since keys/passwords are not stored in the config file, they stick around on the server after we upload them. So if I edit the TC config to add a new password parameter, that password is stored on the server, but outside of the config file. What goes in the config file is a hash which refers back to that password. So when I revert the config changes on the server, that doesn't actually delete the password from the server. It just deletes the configuration change I made. So when I add the correct hash value to the configuration _in VCS_, that hash value still references the password on the server.
- When someone makes a change in the TeamCity GUI, TeamCity attempts to commit that back to `wp-calypso`, which obviously fails. However, since it fails, the TeamCity server gets placed in a state of thinking that its own config file has been modified, but not applied to VCS -- even if the same changes get made to the VCS config file. To resolve this, we have to remove and then re-add the VCS source in TeamCity, thus making TeamCity "restart" with the config from VCS, _or_ we have to revert the change in TeamCity so that it no longer thinks that its own config file is in a "modified" state.

I've run into this problem in the past, and it's fairly tricky. In an ideal world, I would love to disable GUI modifications in TeamCity. However, that makes it very difficult, if not impossible to change certain things. For example, we sometimes make a modification in the GUI, and then copy the code it outputs to VCS without saving the change in the GUI. It also appears impossible to upload a password without modifying the parameters in the GUI. So that at least needs to be accessible for this process to work.

worldomonation commented:
> cc @worldomonation @WunderBart
> 
> We (myself and @griffbrad) were thinking that it would be helpful to allow rotating the config key without causing all existing builds to fail. (Which means everyone has to rebase, and it's hard to communicate that to everyone.) What do you think?

Hey! Sorry for coming in a bit late. 

I think the idea is great and I fully support, but it looks like you've run into issues with the way TeamCity does things.

> I'm surprised this didn't cause issues on previous updates to this file

To my knowledge, modification to the secrets file password hasn't caused the dreaded `not committed to VCS` issue before. I recall someone once accidentally performed a search engine query with the secrets file password prompting a quick change of the said password. It sure causes the builds to fail until rebased, but IIRC `trunk` kept humming along fine.

noahtallen commented:
> modification to the secrets file password hasn't caused the dreaded not committed to VCS issue before

Cool! I think it happens in this case, because we're making a new key instead of changing an existing one. :thinking: 

Either way, I'll be sure to test changing the e2e key after merging this so that we can fully document what happens.

noahtallen committed:
Use previous encryption

noahtallen committed:
Clarify comment
